data/src/main/java/com/linkedin/data/collections/CheckedMap.java

```diff
@@ -356,11 +358,21 @@ public final void addChangeListener(ChangeListener<K, V> listener)
       // Change listeners are mostly used by map wrappers, and we always iterate through them
       // linearly, so use a linked list.
       _changeListeners = new LinkedList<>();
+      _changeListenerReferenceQueue = new ReferenceQueue<>();
```

Comment:
ReferenceQueue maybe fairly heavy. Have we considered a design of having a static global referencequeue and using a cleanup thread to poll and cleanup references? This is similar to what guava does for it's cache with weak keys.

Also if the queue grows long, the iteration through the linkedlist to remove the node will be an O(N) operation everytime. Have we considered using a custom LinkedList impl where we have access to the Node container so that we can do an O(1) removal?

Reply:
Running a separate thread would locking to ensure no race condition when the change listener list is modified. Also, the change listener list is already being updated in multiple threads when the same datamap is wrapped into records in different threads. So I'm worried the O(1) removal logic would run into concurrency issues. 

I'm not that worried about the list growing too big as the change listeners are mostly short-lived (request scoped).

Reply:
To decide between using a doubly linked list and purge as we poll from reference queue vs current impl which traverses the singly linked list separately:

The trade off is extra memory for prev pointer vs additional iteration of the list. And if we use both prev and next pointer, two links are updated for each node during purge.
 So overall the amount of time would be similar: one iteration with two pointers vs two iteration with one pointer.

If we assume the list would be small or most of the items in the list will be purged(this is the expected behavior), the amortized cost of extra iteration is not bad. It gets bad only if a small subset of items are purged.